Update README.md with a simple example (#1529)
cocotb/cocotb: README.md

  * [Join the Gitter chat room](https://gitter.im/cocotb)

## Installation

> ## Requirements

Author: ```suggestion
**Requirements**
```

- Python 3.5+
- A C++ compiler
- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

For more details see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).

> Cocotb can be installed by running `pip install cocotb`.

Reviewer: Or maybe:

```suggestion
Cocotb can be installed by running `pip install cocotb`, and requires:

- Python 3.5+
- A C++ compiler
- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

For more details see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).
```

Author: 🥇

## Usage

For a D flip-flop description:

```verilog
module dff (
    output reg q,
    input c, d,
);

always @(posedge clk) begin
    q <= d;
end

endmodule
```

An example of a simple randomized cocotb testbench:

```python
import random
import cocotb
from cocotb.clock import Clock
from cocotb.triggers import FallingEdge

@cocotb.test()
async def test_simple_dff(dut):
    """ Test that D propagates to Q """
    cocotb.fork(Clock(dut.c, 10, units="us").start())  # Create a 10us period clock on port `c`

    for i in range(10):
        val = random.randint(0, 1)
        dut.d <= val
        await FallingEdge(dut.c)
        assert dut.q == val, "output was incorrect on the {}th cycle".format(i)
```

For more information please see the [cocotb documentation](https://cocotb.readthedocs.io/) and [wiki](https://github.com/cocotb/cocotb/wiki).

## Tutorials, examples and related projects

* [Endian Swapper tutorial](https://cocotb.readthedocs.org/en/latest/endian_swapper.html)
* [Ping using TUN/TAP tutorial](https://cocotb.readthedocs.org/en/latest/ping_tun_tap.html)

> * [Cocotb based USB 1.1 test suite for FPGA IP, with testbenches for a variety of open source USB cores](https://github.com/antmicro/usb-test-suite-build)

Reviewer 2: Maybe just point to https://github.com/cocotb/cocotb/wiki/Further-Resources#libraries-and-examples (and structure that a bit better)?

Author: See below.

> * [Functional Coverage and Constrained Randomization Extensions for Cocotb](https://github.com/mciepluc/cocotb-coveragel)

Reviewer 2: ```suggestion
* [Functional Coverage and Constrained Randomization Extensions for Cocotb](https://github.com/mciepluc/cocotb-coverage)
```

> * [UVM 1.2 port to Python](https://github.com/tpoikela/uvm-python)

Author: ```suggestion
* [UVM 1.2 port to Python](https://github.com/tpoikela/uvm-python)

For more related resources please check the [wiki](https://github.com/cocotb/cocotb/wiki/Further-Resources) and [dependecies](https://github.com/cocotb/cocotb/network/dependents).
```

@cmarqu something like this?

Reviewer 2: Looks good to me if you replace ``dependecies``(typo :)) with ``list of projects depending on cocotb``.

Author: ok now?

Reviewer 2: One "the" after the "and" please, sorry for not spotting this earlier.

  * [Join the Gitter chat room](https://gitter.im/cocotb)

## Installation
Cocotb can be installed by running `pip install cocotb`, and requires:

- Python 3.5+
- A C++ compiler
- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

For more details see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).

## Usage

For a D flip-flop description:

```verilog
module dff (
    output reg q,

>     input c, d,

Author: ```suggestion
    input wire c, d,
```

from cocotb.triggers import FallingEdge

@cocotb.test()
async def test_simple_dff(dut):
    """ Test that D propagates to Q """
    
    clock = Clock(dut.c, 10, units="us")  # Create a 10us period clock on port `c`
    cocotb.fork(clock.start())  # Start the clock

    for i in range(10):
        val = random.randint(0, 1)
        dut.d <= val
        await FallingEdge(dut.c)
        assert dut.q == val, "output was incorrect on the {}th cycle".format(i)
```

For more information please see the [cocotb documentation](https://cocotb.readthedocs.io/) and [wiki](https://github.com/cocotb/cocotb/wiki).

## Tutorials, examples and related projects

* [Endian Swapper tutorial](https://cocotb.readthedocs.org/en/latest/endian_swapper.html)
* [Ping using TUN/TAP tutorial](https://cocotb.readthedocs.org/en/latest/ping_tun_tap.html)
* [Cocotb based USB 1.1 test suite for FPGA IP, with testbenches for a variety of open source USB cores](https://github.com/antmicro/usb-test-suite-build)
* [Functional Coverage and Constrained Randomization Extensions for Cocotb](https://github.com/mciepluc/cocotb-coverage)
* [UVM 1.2 port to Python](https://github.com/tpoikela/uvm-python)

> For more related resources please check the [wiki](https://github.com/cocotb/cocotb/wiki/Further-Resources) and [list of projects depending on cocotb](https://github.com/cocotb/cocotb/network/dependents).

Author: ```suggestion
For more related resources please check the [wiki](https://github.com/cocotb/cocotb/wiki/Further-Resources) and the [list of projects depending on cocotb](https://github.com/cocotb/cocotb/network/dependents).
```

Cocotb can be installed by running `pip install cocotb`.

## Usage

For a D flip-flop description: 

```verilog
module dff (
    output reg q,
    input c, d,
);

always @(posedge clk) begin
    q <= d;
end

endmodule
```

> An example of simple randomized Cocotb based testbanch:

Reviewer: ```suggestion
An example of simple randomized Cocotb based testbench:
```

```python
import random
import cocotb
from cocotb.clock import Clock
from cocotb.triggers import FallingEdge

@cocotb.test()

> async def test_simple_dff(dut):

Reviewer: ```suggestion
async def test_simple_dff(dut):
    """ Test that D propagates to Q """
```

>     cocotb.fork(Clock(dut.c, 10, "us").start()) # Create a 10us period clock on port `c`

Reviewer: ```suggestion
    cocotb.fork(Clock(dut.c, 10, units="us").start())  # Create a 10us period clock on port `c`
```

> 

Author: Maybe ?
```suggestion
In order to use cocotb Python interpreter, c/c++ compiler and HDL simulator are required to see more in [documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites)
```

Reviewer: I thought your earlier comment was pretty spot on:
```suggestion
## Requirements
- Python 3.5+
- A C++ compiler
- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

For more details see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).
```

Cocotb can be installed by running `pip install cocotb`.

## Usage

> 

Reviewer: Do we want to keep a mention of iverilog somewhere?

Author: Good question we could add something like:

### Prerequisites
- Python 3.5+ (with --embedded-shared)
- a compiler gcc/clang 
- an HDL simulator (example [Icarus Verilog](http://iverilog.icarus.com/)

Reviewer: That seems pretty reasonable

Reviewer 3: I'd remove the `--embedded-shared` part. This is a quickstart README, optimized for the standard user, who will get Python from a package manager. Instead, link to more detailed requirements (incl. compiler requirements) in the docs. 

Also link to a list of supported simulators, I get that question regularly.

Author: Maybe something like this:

In order to use cococtb Python interpreter, c/c++ compiler and HDL simulator is required to see more in [coumenation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites)

To not double the same information that can change.

Reviewer: Agreed that we don't want the list in the readme. I think it's ok to name examples of such a simulator in the readme though.

For a D flip-flop description:

```verilog
module dff (
    output reg q,
    input c, d,
);

always @(posedge clk) begin
    q <= d;
end

endmodule
```

> An example of simple randomized Cocotb based testbench:

Reviewer 2: ```suggestion
An example of a simple randomized cocotb testbench:
```

```python
import random
import cocotb
from cocotb.clock import Clock
from cocotb.triggers import FallingEdge

@cocotb.test()
async def test_simple_dff(dut):
    """ Test that D propagates to Q """

>     cocotb.fork(Clock(dut.c, 10, units="us").start())  # Create a 10us period clock on port `c`

Author: ```suggestion
    
    clock = Clock(dut.c, 10, units="us") # Create a 10us period clock on port `c`
    cocotb.fork(clock.start()) # Start clock
```
Maybe?

Reviewer 2: PEP-8 says to use double space before comments.

    for i in range(10):
        val = random.randint(0, 1)
        dut.d <= val
        await FallingEdge(dut.c)
        assert dut.q == val, "output was incorrect on the {}th cycle".format(i)
```

> For more information please see [Cocotb documentation](https://cocotb.readthedocs.io/) and [wiki](https://github.com/cocotb/cocotb/wiki).

Reviewer 2: ```suggestion
For more information please see the [cocotb documentation](https://cocotb.readthedocs.io/) and [wiki](https://github.com/cocotb/cocotb/wiki).
```

  * [Join the Gitter chat room](https://gitter.im/cocotb)

## Installation

> Cocotb can be installed by running `pip install cocotb`, and requires:

Author: Can change. Mind python3 is forced in `setup.py`.

> Cocotb can be installed by running `pip install cocotb`, and requires:

Author: You won't get pip3 with conda. Maybe `pip[3]` ?

> Cocotb can be installed by running `pip install cocotb`, and requires:

Reviewer 3: `python3 -m pip` maybe?

> Cocotb can be installed by running `pip install cocotb`, and requires:

Reviewer: It's worth noting the instructions on PyPI are `pip install cocotb`: https://pypi.org/project/cocotb/

> Cocotb can be installed by running `pip install cocotb`, and requires:

Reviewer 3: How do other projects handle that? `pip install` on almost all "stable" Linux distributions today gives you Python 2, and errors out during the installation. I know many "beginners" (and in the EDA world, many people are "Python beginners") have stumbled across that, together with the need for `--user` in many cases. [Insert rant about Python packaging usability here.]

Maybe in this PR:
- Leave as-is with `pip`

In follow-ups:
- Improve the RuntimeException in setup.py to mention something like "have you tried pip3 instead of pip"?
- Add OS-specific install instructions, which also cover the compiler, etc. in the docs, and link to them in the README once we have them.

> Cocotb can be installed by running `pip install cocotb`, and requires:

Reviewer: Picking a large non-binary framework at random, django recommends the simple `pip` invocation [here](https://www.djangoproject.com/download/), and the slightly longer `python -m pip` invocation [here](https://docs.djangoproject.com/en/3.0/topics/install/#installing-official-release).

Perhaps linking to `pip` would help - I don't really want our readme to explain exactly how users should be using `pip` on their system, that feels like either `pip` or the user's problem.

You could consider opening an issue against https://github.com/pypa/pypi-support, and ask what they recommend give the top left of the package pages uses `pip install <...>`.

> Cocotb can be installed by running `pip install cocotb`, and requires:

Reviewer 3: https://github.com/cocotb/cocotb/pull/1558 should at least improve things a bit in a true Python way ("Easier to ask for forgiveness than permission.").

- Python 3.5+
- A C++ compiler

> - An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

Reviewer 3: Please turn things around here and put them in "action" style: first the requirements, then the installation.

> For more details see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).

Reviewer 3: ```suggestion
For more details, including how to install a development version of cocotb, see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).
```

(This README is also the entry page of the pypi listing.)

## Usage

> 

Reviewer 3: Start with a sentence of what you're trying to show here, e.g.

As first trivial introduction to cocotb, the following example "tests" a flip-flop.

For a D flip-flop description:

> ```verilog

Reviewer 3: People often fail to understand what to do with those code snippets; add a comment at the beginning with a file name to give a hint:

```
# dff.sv
```

module dff (
    output reg q,

>     input wire c, d,

Reviewer 3: Please use `logic`, we don't want to look like a Verilog code base from the 90s.

Author: OK for me but it looks that Icarus 10.3 will not handle this.

>     input wire c, d,

Reviewer 3: c -> clk to match the RTL below.

);

> always @(posedge clk) begin

Reviewer 3: always -> always_ff

Author: OK for me but it looks that Icarus 10.3 will not handle this.

    q <= d;
end

endmodule
```

An example of a simple randomized cocotb testbench:

> ```python

Reviewer 3: Add 
```
# test_dff.py
```

import random
import cocotb
from cocotb.clock import Clock
from cocotb.triggers import FallingEdge

@cocotb.test()

> async def test_simple_dff(dut):

Reviewer 3: suggestion: `test_dff_simple()` (filename, then test name)

    """ Test that D propagates to Q """
    

>     clock = Clock(dut.c, 10, units="us")  # Create a 10us period clock on port `c`

Reviewer 3: c -> clk to match the RTL (and "c" isn't really a common name for a clock in RTL)

    cocotb.fork(clock.start())  # Start the clock

    for i in range(10):
        val = random.randint(0, 1)
        dut.d <= val
        await FallingEdge(dut.c)
        assert dut.q == val, "output was incorrect on the {}th cycle".format(i)
```

> 

Reviewer 3: To answer the question "what to do with these pieces of code": 
- Add the most trivial makefile for icarus, and a simple command line invocation with the output of cocotb, e.g.

```
$ make SIM=icarus
...
some log output ...
```

Bonus points for an asciinema version :smile:

## Installation

Cocotb requires:

- Python 3.5+
- A C++ compiler
- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

and can be installed by running `pip install cocotb`. 
For more details, including how to install a development version of cocotb, see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).

## Usage

For a D flip-flop description:

```verilog
// dff.v

module dff (
    output reg q,
    input wire clk, d
);

always @(posedge clk) begin
    q <= d;
end

endmodule
```

An example of a simple randomized cocotb testbench:

```python
# test_dff.py

import random
import cocotb
from cocotb.clock import Clock
from cocotb.triggers import FallingEdge

@cocotb.test()
async def test_dff_simple(dut):
    """ Test that D propagates to Q """

>     clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port `c`

Reviewer 2: ```suggestion
    clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port `clk`
```

## Installation

Cocotb requires:

- Python 3.5+
- A C++ compiler
- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

and can be installed by running `pip install cocotb`.
For more details, including how to install a development version of cocotb, see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).

## Usage

As first trivial introduction to cocotb, the following example "tests" of a flip-flop.

For a D flip-flop description:

```verilog
// dff.v

`timescale 1us/1ns

module dff (
    output reg q,
    input wire clk, d
);

always @(posedge clk) begin
    q <= d;
end

endmodule
```

An example of a simple randomized cocotb testbench:

```python
# test_dff.py

import random
import cocotb
from cocotb.clock import Clock
from cocotb.triggers import FallingEdge

@cocotb.test()
async def test_dff_simple(dut):
    """ Test that D propagates to Q """

    clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port `clk`
    cocotb.fork(clock.start())  # Start the clock

    for i in range(10):
        val = random.randint(0, 1)

>         dut.d <= val

Reviewer 2: Indeed, someone on Gitter was just asking what this syntax means.

        await FallingEdge(dut.clk)
        assert dut.q == val, "output was incorrect on the {}th cycle".format(i)
```

A simple Makefile:

```make
# Makefile

TOPLEVEL_LANG=verilog
VERILOG_SOURCES=$(shell pwd)/dff.v
MODULE := test_dff
TOPLEVEL = dff

include $(shell cocotb-config --makefiles)/Makefile.inc
include $(shell cocotb-config --makefiles)/Makefile.sim
```

To run the test execute:

```bash
SIM=icarus make
```

> For more information please see the [cocotb documentation](https://cocotb.readthedocs.io/) and [wiki](https://github.com/cocotb/cocotb/wiki).

Reviewer 3: ```suggestion
For more information please see the [cocotb documentation](https://cocotb.readthedocs.io/) and the [wiki](https://github.com/cocotb/cocotb/wiki).
```

## Installation

Cocotb requires:

- Python 3.5+
- A C++ compiler
- An HDL simulator (such as [Icarus Verilog](http://iverilog.icarus.com/))

After installing these dependencies, the latest stable version of cocotb can be installed with pip.

```command
pip install cocotb
```

> For more details, including how to install a development version of cocotb, see [the documentation](https://cocotb.readthedocs.io/en/latest/quickstart.html#pre-requisites).

Reviewer 3: ```suggestion
For more details, including how to install a development version of cocotb, see [the documentation](https://docs.cocotb.org/en/latest/quickstart.html#pre-requisites).
```

## Usage

> As first trivial introduction to cocotb, the following example "tests" a flip-flop.

Reviewer 4: ```suggestion
As a first trivial introduction to cocotb, the following example "tests" a flip-flop.
```

First, we need a hardware design which we can test. For this example, create a file `dff.sv` with SystemVerilog code for a simple [D flip-flop](https://en.wikipedia.org/wiki/Flip-flop_(electronics)#D_flip-flop). You could also use any other language a [cocotb-supported simulator](https://cocotb.readthedocs.io/en/latest/introduction.html) understands, e.g. VHDL.

> ```verilog

Reviewer 3: verilog -> systemverilog for full syntax highlighting.

// dff.sv

`timescale 1us/1ns

module dff (
    output logic q,
    input logic clk, d
);

always @(posedge clk) begin
    q <= d;
end

endmodule
```

An example of a simple randomized cocotb testbench:

```python
# test_dff.py

import random
import cocotb
from cocotb.clock import Clock
from cocotb.triggers import FallingEdge

@cocotb.test()
async def test_dff_simple(dut):
    """ Test that d propagates to q """

    clock = Clock(dut.clk, 10, units="us")  # Create a 10us period clock on port clk
    cocotb.fork(clock.start())  # Start the clock

    for i in range(10):

>         val = random.randint(0, 1)

Reviewer 4: It's good practice. You would be surprised how many bugs I've seen because people have assumed "when" they are.
```suggestion
        await RisingEdge(dut.clk)  # Synchronize ourselves with the clock
        val = random.randint(0, 1)
```

Author: In this form, the test fails. `ValueError: Unable to resolve to binary >z<` . And I am not sure in any case.

Author: ```
  6000.00ns ERROR    cocotb.regression                         regression.py:396  in _score_test                     Test Failed: test_dff_simple (result was ValueError)
                                                                                                                     Traceback (most recent call last):
                                                                                                                       File "/examples/dff_simple/test_dff.py", line 20, in test_dff_simple
                                                                                                                         assert dut.q == val, "output was incorrect on the {}th cycle".format(i)
                                                                                                                       File "/cocotb/handle.py", line 409, in __eq__
                                                                                                                         return self.value == other
                                                                                                                       File "/cocotb/binary.py", line 439, in __eq__
                                                                                                                         return self.value == other
                                                                                                                       File "/cocotb/binary.py", line 296, in value
                                                                                                                         return self.integer
                                                                                                                       File "/cocotb/binary.py", line 287, in integer
                                                                                                                         return self._convert_from[self.binaryRepresentation](self._str)
                                                                                                                       File "/cocotb/binary.py", line 193, in _convert_from_unsigned
                                                                                                                         return int(resolve(x), 2)
                                                                                                                       File "/cocotb/binary.py", line 45, in resolve
                                                                                                                         raise ValueError("Unable to resolve to binary >%s<" % string)
                                                                                                                     ValueError: Unable to resolve to binary >z<
  6000.00ns ERROR    cocotb.regression                         regression.py:227  in tear_down                       Failed 1 out of 1 tests (0 skipped)
```

Reviewer 4: The clock hasn't gotten a value yet. This is really more of an issue of `Clock`, which is essentially a driver (just *not*), not running `setimmediate` on the clock signal.

This also highlights my absolute biggest peeve of all about cocotb. The stacktrace there is completely useless. And going in the other direction I sometimes see `truncating bits to match required width` warning messages, but it doesn't tell me on what handle and with what value I tried to do that, or where in my program I am. Sorry to rant in your PR.

Reviewer: > The stacktrace there is completely useless

I don't think it's all that bad. It's not too hard (for me) to infer that `dut.q.__eq__` is failing, and that it's because `q` contains a `z`. Perhaps come comments at the end of the lines to aid the user would help, but I don't think the stacktrace is actually missing much information.

Reviewer 4: Hmmm, I just glanced at it, that's different than what we get, but we are pretty far behind. I guess this was recently fixed.

Reviewer 2: "doesn't tell me on what handle and with what value I tried to do that" - at least printing the value would be trivial to add.

Author: Reported issue: https://github.com/cocotb/cocotb/issues/1611

        dut.d <= val  # Assign the random value val to the input port d
        await FallingEdge(dut.clk)
        assert dut.q == val, "output q was incorrect on the {}th cycle".format(i)
```

A simple Makefile:

```make
# Makefile

TOPLEVEL_LANG = verilog
VERILOG_SOURCES = $(shell pwd)/dff.sv
TOPLEVEL = dff
MODULE = test_dff

include $(shell cocotb-config --makefiles)/Makefile.inc
include $(shell cocotb-config --makefiles)/Makefile.sim
```

> In order to run the test, execute:

Reviewer 3: ```suggestion
In order to run the test with Icarus Verilog, execute:
```

```command
make SIM=icarus
```

> [![asciicast](https://asciinema.org/a/316949.svg)](https://asciinema.org/a/316949)

Reviewer 3: Great, thanks! :rocket:

> [![asciicast](https://asciinema.org/a/316949.svg)](https://asciinema.org/a/316949)

Reviewer: Do we want to rebuild this now that #1581 is in, or do we not care?

Author: Heh. If you really want. Need some ❤️

Reviewer: Could edit with https://discourse.asciinema.org/t/can-i-edit-post-process-the-recording/25

> For more information please see the [cocotb documentation](https://cocotb.readthedocs.io/) and the [wiki](https://github.com/cocotb/cocotb/wiki).

Reviewer 3: Switched to docs.cocotb.org now
